Code/src/main/java/nl/utwente/viskell/ui/components/FunApplyBlock.java

```diff
@@ -0,0 +1,224 @@
+package nl.utwente.viskell.ui.components;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+
+import javafx.fxml.FXML;
+import javafx.geometry.BoundingBox;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.layout.HBox;
+import javafx.scene.layout.Pane;
+import javafx.scene.layout.VBox;
+import nl.utwente.viskell.haskell.env.FunctionInfo;
+import nl.utwente.viskell.haskell.expr.Apply;
+import nl.utwente.viskell.haskell.expr.Binder;
+import nl.utwente.viskell.haskell.expr.Expression;
+import nl.utwente.viskell.haskell.expr.FunVar;
+import nl.utwente.viskell.haskell.expr.Lambda;
+import nl.utwente.viskell.haskell.expr.LocalVar;
+import nl.utwente.viskell.haskell.type.FunType;
+import nl.utwente.viskell.haskell.type.Type;
+import nl.utwente.viskell.haskell.type.TypeScope;
+import nl.utwente.viskell.ui.CustomUIPane;
+import nl.utwente.viskell.ui.DragContext;
+
+public class FunApplyBlock extends Block {
+    
+    /** Function input anchor that can be dragged down for currying. */
+    private class FunInputAnchor extends Pane {
+
+        /** The connection anchor for this input argument. */
+        private final InputAnchor anchor;
+        
+        /** The draggable input type label. */
+        private final Label inputType;
+
+        /** Whether this input argument is in curried state. */
+        private boolean curried;
+        
+        /** The drag handler for this. */
+        private final DragContext dragContext;
+        
+        public FunInputAnchor() {
+            this.curried = false;
+            this.inputType = new Label(".....") {
+                @Override
+                public void relocate(double x, double y) {
+                    super.relocate(x, y);
+                    FunInputAnchor.this.dragShift(y);
+                }
+            };
+            this.inputType.setMinWidth(USE_PREF_SIZE);
+            this.inputType.getStyleClass().add("inputType");
+            this.anchor = new InputAnchor(FunApplyBlock.this);
+            this.anchor.layoutXProperty().bind(this.inputType.widthProperty().divide(2));
```

Comment:
I'll allow using a bind here - it's simple enough and there's only one of them - but officially prefer getting it right with layouts.